src/chessboard/components/Piece.tsx

```diff
@@ -103,14 +104,13 @@ export function Piece({
   useEffect(() => {
     const removedPiece = positionDifferences.removed?.[square];
     // return as null and not loaded yet
-    if (!positionDifferences.added) return;
+    if (!positionDifferences.added || !removedPiece) return;
     // check if piece matches or if removed piece was a pawn and new square is on 1st or 8th rank (promotion)
     const newSquare = (
       Object.entries(positionDifferences.added) as [Square, Pc][]
     ).find(
       ([s, p]) =>
-        p === removedPiece ||
-        (removedPiece?.[1] === "P" && (s[1] === "1" || s[1] === "8"))
+        p === removedPiece || onPromotionCheck(square, s, removedPiece)
```

Comment:
before 

https://github.com/Clariity/react-chessboard/assets/42891870/2bd777d7-9eab-42d0-a503-2c50d96a4a81







and after 

https://github.com/Clariity/react-chessboard/assets/42891870/b1e76105-699e-4367-93a1-08759aec8f1d